processing/src/main/java/org/apache/druid/math/expr/ConstantExpr.java

```diff
@@ -39,9 +40,12 @@
  * {@link Expr.ObjectBinding}. {@link ConstantExpr} are terminal nodes of an expression tree, and have no children
  * {@link Expr}.
  */
+@Immutable
 abstract class ConstantExpr<T> implements Expr
 {
   final ExpressionType outputType;
+
+  @SuppressWarnings("Immutable")
```

Comment:
What does `@Immutable` do and why are these various `@SuppressWarnings("Immutable")` necessary (here, in BaseTypeSignature, in ColumnCapabilitiesImpl)?

I am not familiar with this annotation so was just wondering what the extra lines are for.

Reply:
* the `@Immutable` enables a compile time check that the [class satisfies some immutable rules](https://errorprone.info/bugpattern/Immutable)
   * such a check could have warned me before doing the original PR
* the suppression is added for this field as we don't know what it is...but since this class will handle it as a blackbox I think its okay to ignore it
* I had to also add `SuppressWarnings("Immutable")` to `ColumnCapabiliiesImpl` because its kinda immutable; but fixing that turned into a massacre - I've asked @clintropolis  about the value of such a change and based on that conversion I opted to not touch a lot of files without much benefits

should I remove these `@Immutable` related changes?

Reply:
Ah, it's an errorprone thing? Cool. I was just wondering. I don't think we should necessarily aggressively use it everywhere (we have a lot of immutable classes, seems like it would be a lot of annotation noise) but it seems useful here.